crates/win-api-wrappers/src/wts.rs

```diff
@@ -87,24 +98,223 @@ impl Drop for WTSVirtualChannel {
     }
 }
 
+#[repr(i32)]
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum WTSConnectState {
+    Active = windows::Win32::System::RemoteDesktop::WTSActive.0,
+    Connected = windows::Win32::System::RemoteDesktop::WTSConnected.0,
+    ConnectQuery = windows::Win32::System::RemoteDesktop::WTSConnectQuery.0,
+    Shadow = windows::Win32::System::RemoteDesktop::WTSShadow.0,
+    Disconnected = windows::Win32::System::RemoteDesktop::WTSDisconnected.0,
+    Idle = windows::Win32::System::RemoteDesktop::WTSIdle.0,
+    Listen = windows::Win32::System::RemoteDesktop::WTSListen.0,
+    Reset = windows::Win32::System::RemoteDesktop::WTSReset.0,
+    Down = windows::Win32::System::RemoteDesktop::WTSDown.0,
+    Init = windows::Win32::System::RemoteDesktop::WTSInit.0,
+}
+
+impl TryFrom<WTS_CONNECTSTATE_CLASS> for WTSConnectState {
+    type Error = anyhow::Error;
+
+    fn try_from(v: WTS_CONNECTSTATE_CLASS) -> Result<Self, Self::Error> {
+        match v {
+            windows::Win32::System::RemoteDesktop::WTSActive => Ok(WTSConnectState::Active),
+            windows::Win32::System::RemoteDesktop::WTSConnected => Ok(WTSConnectState::Connected),
+            windows::Win32::System::RemoteDesktop::WTSConnectQuery => Ok(WTSConnectState::ConnectQuery),
+            windows::Win32::System::RemoteDesktop::WTSShadow => Ok(WTSConnectState::Shadow),
+            windows::Win32::System::RemoteDesktop::WTSDisconnected => Ok(WTSConnectState::Disconnected),
+            windows::Win32::System::RemoteDesktop::WTSIdle => Ok(WTSConnectState::Idle),
+            windows::Win32::System::RemoteDesktop::WTSListen => Ok(WTSConnectState::Listen),
+            windows::Win32::System::RemoteDesktop::WTSReset => Ok(WTSConnectState::Reset),
+            windows::Win32::System::RemoteDesktop::WTSDown => Ok(WTSConnectState::Down),
+            windows::Win32::System::RemoteDesktop::WTSInit => Ok(WTSConnectState::Init),
+            _ => Err(anyhow::anyhow!("Invalid WTS_CONNECTSTATE_CLASS: {}", v.0)),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct WTSSessionInfo {
+    pub session_id: u32,
+    pub win_station_name: String,
+    pub state: WTSConnectState,
+}
+
+impl TryFrom<&WTS_SESSION_INFOW> for WTSSessionInfo {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &WTS_SESSION_INFOW) -> Result<Self, Self::Error> {
+        let win_station_name: String;
+
+        // SAFETY: `pWinStationName` will always hold a valid WSTR pointer if
+        // `WTS_SESSION_INFOW` was returned from Remote Desktop Services (`WTSEnumerateSessionsW``)
+        unsafe { win_station_name = value.pWinStationName.to_string()? }
```

Comment:
issue: It’s possible to construct a WTS_SESSION_INFOW in safe Rust, and to call this function like `WTSSessionInfo::try_from(my_session_info)`, and this may cause memory corruptions.

And it’s technically okay to create a WTS_SESSION_INFOW on Rust side. The precondition is not really about being allocated my Remote Desktop Services. It’s more about "the pointer being valid for reads up until and including the next `\0`": https://docs.rs/windows-strings/latest/windows_strings/struct.PWSTR.html#method.as_wide

This is called an "soundness" bug and the API is said "unsound": https://docs.rs/dtolnay/latest/dtolnay/macro._03__soundness_bugs.html

suggestion: Remove this `TryFrom` implementation, and replace it by a `unsafe fn new(&WTS_SESSION_INFOW) { … }` function with a `# Safety` section basically requiring the same preconditions as `PWSTR::as_wide()`.

Reply:
Indeed; my thinking here was that since this is a private function we _know_ we're not building `WTS_SESSION_INFOW` ourselves. Since we _could_ do this, however, I understand the refactoring. Starting to think more like a rustacean :)

Reply:
I would avoid this kind of implicit contracts even with a private function, but keep in mind that the functions implemented via a trait have the same visibility as the trait, i.e.: for TryFrom it’s public. So when I reviewed I considered this as an unsound _public_ API :slightly_smiling_face: